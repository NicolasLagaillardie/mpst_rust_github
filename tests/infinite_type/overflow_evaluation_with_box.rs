use crossbeam_channel::{Receiver, Sender};
use std::marker;
// use std::rc::Rc;

// Trait
pub trait Session: marker::Sized + marker::Send {
    type Dual: Session<Dual = Self>;
}

// Struct
#[derive(Debug)]
pub struct Send<T, S>
where
    T: marker::Send,
    S: Session,
    S::Dual: Session,
{
    pub channel: Sender<(T, S::Dual)>,
}

#[derive(Debug)]
pub struct Recv<T, S>
where
    T: marker::Send,
    S: Session,
{
    pub channel: Receiver<(T, S)>,
}

#[derive(Debug)]
pub struct End {}

// impl
impl<T: marker::Send, S: Session> Session for Send<T, S> {
    type Dual = Recv<T, S::Dual>;
}

impl<T: marker::Send, S: Session> Session for Recv<T, S> {
    type Dual = Send<T, S::Dual>;
}

impl Session for End {
    type Dual = End;
}

// enum
enum Test0 {
    End(Box<End>),
    Branch(Box<Recv<Test0, End>>),
}
enum Test1 {
    End(Box<End>),
    Branch(Box<Recv<i32, Recv<Test1, End>>>),
}
enum Test2 {
    End(Box<End>),
    Branch(Box<Recv<i32, Send<Test2, End>>>),
}
enum Test3 {
    End(Box<End>),
    Branch(Box<Send<i32, Recv<Test3, End>>>),
}
enum Test4 {
    End(Box<End>),
    Branch(Box<Send<i32, Send<Test4, End>>>),
}
enum Test5 {
    End(Box<End>),
    Branch(Box<Recv<i32, Recv<i32, Recv<Test5, End>>>>),
}
enum Test6 {
    End(Box<End>),
    Branch(Box<Recv<i32, Recv<i32, Send<Test6, End>>>>),
}
enum Test7 {
    End(Box<End>),
    Branch(Box<Recv<i32, Send<i32, Recv<Test7, End>>>>),
}
enum Test8 {
    End(Box<End>),
    Branch(Box<Recv<i32, Send<i32, Send<Test8, End>>>>),
}
enum Test9 {
    End(Box<End>),
    Branch(Box<Send<i32, Recv<i32, Recv<Test9, End>>>>),
}
enum Test10 {
    End(Box<End>),
    Branch(Box<Send<i32, Recv<i32, Send<Test10, End>>>>),
}
enum Test11 {
    End(Box<End>),
    Branch(Box<Send<i32, Send<i32, Recv<Test11, End>>>>),
}
enum Test12 {
    End(Box<End>),
    Branch(Box<Send<i32, Send<i32, Send<Test12, End>>>>),
}

fn main() {}
