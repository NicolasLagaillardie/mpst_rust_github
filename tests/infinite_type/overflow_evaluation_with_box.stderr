error[E0275]: overflow evaluating the requirement `Send<i32, Recv<i32, Recv<Test9, End>>>: std::marker::Send`
  --> tests/infinite_type/overflow_evaluation_with_box.rs:85:12
   |
85 |     Branch(Box<Send<i32, Recv<i32, Recv<Test9, End>>>>),
   |            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
   |
   = note: required for `Unique<Send<i32, Recv<i32, Recv<Test9, End>>>>` to implement `std::marker::Send`
note: required because it appears within the type `Box<Send<i32, Recv<i32, Recv<Test9, End>>>>`
  --> $RUST/alloc/src/boxed.rs
   |
   | pub struct Box<
   |            ^^^
note: required because it appears within the type `Test9`
  --> tests/infinite_type/overflow_evaluation_with_box.rs:83:6
   |
83 | enum Test9 {
   |      ^^^^^
note: required for `Recv<Test9, End>` to implement `Session`
  --> tests/infinite_type/overflow_evaluation_with_box.rs:38:35
   |
38 | impl<T: marker::Send, S: Session> Session for Recv<T, S> {
   |         ------------              ^^^^^^^     ^^^^^^^^^^
   |         |
   |         unsatisfied trait bound introduced here

error[E0275]: overflow evaluating the requirement `Send<i32, Recv<i32, Send<Test10, End>>>: std::marker::Send`
  --> tests/infinite_type/overflow_evaluation_with_box.rs:89:12
   |
89 |     Branch(Box<Send<i32, Recv<i32, Send<Test10, End>>>>),
   |            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
   |
   = note: required for `Unique<Send<i32, Recv<i32, Send<Test10, End>>>>` to implement `std::marker::Send`
note: required because it appears within the type `Box<Send<i32, Recv<i32, Send<Test10, End>>>>`
  --> $RUST/alloc/src/boxed.rs
   |
   | pub struct Box<
   |            ^^^
note: required because it appears within the type `Test10`
  --> tests/infinite_type/overflow_evaluation_with_box.rs:87:6
   |
87 | enum Test10 {
   |      ^^^^^^
note: required for `Send<Test10, End>` to implement `Session`
  --> tests/infinite_type/overflow_evaluation_with_box.rs:34:35
   |
34 | impl<T: marker::Send, S: Session> Session for Send<T, S> {
   |         ------------              ^^^^^^^     ^^^^^^^^^^
   |         |
   |         unsatisfied trait bound introduced here

error[E0275]: overflow evaluating the requirement `Send<i32, Send<i32, Recv<Test11, End>>>: std::marker::Send`
  --> tests/infinite_type/overflow_evaluation_with_box.rs:93:12
   |
93 |     Branch(Box<Send<i32, Send<i32, Recv<Test11, End>>>>),
   |            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
   |
   = note: required for `Unique<Send<i32, Send<i32, Recv<Test11, End>>>>` to implement `std::marker::Send`
note: required because it appears within the type `Box<Send<i32, Send<i32, Recv<Test11, End>>>>`
  --> $RUST/alloc/src/boxed.rs
   |
   | pub struct Box<
   |            ^^^
note: required because it appears within the type `Test11`
  --> tests/infinite_type/overflow_evaluation_with_box.rs:91:6
   |
91 | enum Test11 {
   |      ^^^^^^
note: required for `Recv<Test11, End>` to implement `Session`
  --> tests/infinite_type/overflow_evaluation_with_box.rs:38:35
   |
38 | impl<T: marker::Send, S: Session> Session for Recv<T, S> {
   |         ------------              ^^^^^^^     ^^^^^^^^^^
   |         |
   |         unsatisfied trait bound introduced here

error[E0275]: overflow evaluating the requirement `Send<i32, Send<i32, Send<Test12, End>>>: std::marker::Send`
  --> tests/infinite_type/overflow_evaluation_with_box.rs:97:12
   |
97 |     Branch(Box<Send<i32, Send<i32, Send<Test12, End>>>>),
   |            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
   |
   = note: required for `Unique<Send<i32, Send<i32, Send<Test12, End>>>>` to implement `std::marker::Send`
note: required because it appears within the type `Box<Send<i32, Send<i32, Send<Test12, End>>>>`
  --> $RUST/alloc/src/boxed.rs
   |
   | pub struct Box<
   |            ^^^
note: required because it appears within the type `Test12`
  --> tests/infinite_type/overflow_evaluation_with_box.rs:95:6
   |
95 | enum Test12 {
   |      ^^^^^^
note: required for `Send<Test12, End>` to implement `Session`
  --> tests/infinite_type/overflow_evaluation_with_box.rs:34:35
   |
34 | impl<T: marker::Send, S: Session> Session for Send<T, S> {
   |         ------------              ^^^^^^^     ^^^^^^^^^^
   |         |
   |         unsatisfied trait bound introduced here
